### ISSUE TITLE
Duplicate PackageReference includes are handled inconsistently by all restores, leading to unnecessary full restores

### ISSUE THREAD

**Issue by nkolev92** [NuGet/Home#9864]
Take a project such as the following: 

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>net5.0</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="NuGet.Protocol" Version="5.5.0" />
    <PackageReference Include="NuGet.Protocol" Version="5.6.0" />
  </ItemGroup>

</Project>
```

Note the duplicate PackageReference declaration. Digging further reveals that different project types and different tools handle it differently. 

See the below table for details

| Restore Flavor | PackageReference declaration | 
| - | - |
| Commandline restore (all) | First |
| Static graph restore (all) | First | 
| VS - SDK based projects | Last |
| VS - Legacy PR projects | First |

This leads to issues as the ones described in https://github.com/dotnet/arcade/issues/5550 (see below repro). 

I think logically from an msbuild perspective, we'd expect that the last value is the one that's resolved. This needs investigated and ideally consolidated. 
Note that consolidating VS - LEgacy PR projects might be impossible, due to the fact that it's largely in maintenance mode.

See repro below.

cc folks that might have some thoughts around: 

@jeffkl @davkean @dsplaisted @zivkan 

fyi @sharwell @garath

[DuplicateIds.zip](https://github.com/NuGet/Home/files/5002828/DuplicateIds.zip)

**WIP**: 
**Design**: https://github.com/NuGet/Home/compare/dev-nkolev92-duplicatesHandling 
**Implementation**: https://github.com/NuGet/NuGet.Client/tree/dev-nkolev92-duplicatePRs

**Comment by dsplaisted** [comment]
Can we make it an error to have duplicated PackageReference items?  This would break projects, but seems like saner behavior.

The SDK already checks for a variety of duplicate items (Compile, FrameworkReference, etc) and errors if there are duplicates.

**Comment by nkolev92** [comment]
The commandline scenarios would be easy. 

In VS, nomination that'd be problematic. I don't think the project-system can/should surface this error, rather it should be NuGet. That means NuGet needs to receive all the data.

The IVsReferenceItem that represents a PR has a name, which suggests it should be unique. 
However, IVsReferenceItems which is the collection of items doesn't really rely on the name as a unique identifier.

Thoughts @davkean? 

I'd have to investigate legacy to understand what happens there.

**Comment by dsplaisted** [comment]
Can you generate the error in the NuGet targets?  Then you would have access to all the MSBuild items.

**Comment by nkolev92** [comment]
That might work.

**Comment by davkean** [comment]
Make `CollectPackageReferences` produce an error, be sure to respect `$(ContinueOnError)` (https://github.com/dotnet/msbuild/blob/6f41a4fa3c6957d430783b7d31fe16e6a13b6e8e/src/Shared/XMakeAttributes.cs#L69) so we don't break the design-time build.

**Comment by davkean** [comment]
Note, we cannot pass through the entire list of items to NuGet ourselves - the entire surface area in CPS is based on dictionaries and not a fair representation of the "list" concept in MSBuild.

**Comment by nkolev92** [comment]
fyi @rconard @anangaur for awareness.

**Comment by nkolev92** [comment]
I have confirmed at least 2 other bugs to the same root cause. 
I'd expect that at least 2 more might be hitting this. 
no-op is very important for us, especially across same version of commandline + VS restore, bumping the pri per that.

**Comment by nkolev92** [comment]
A precedent very much exists in https://github.com/dotnet/sdk/blob/44f381b62d466565639d51847c9127afbe7062a9/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.DefaultItems.targets#L284 :) 

I think the biggest concern here is how *happy* customers will be to suddenly start getting this error. :)
This is a categorically **good** change though, the behavior across all tools should be consistent.

**Comment by dsplaisted** [comment]
To mitigate the impact of breaking people, I'd suggest a flag they can set to revert to the old behavior (or perhaps to revert to consistent "choose last" or "choose first" behavior across the four different restore types).

**Comment by marcin-krystianc** [comment]
> I think the biggest concern here is how happy customers will be to suddenly start getting this error. :)
This is a categorically good change though, the behavior across all tools should be consistent.

I lost in the past quite a chunk of time trying to investigate issues caused by duplicated `PackageReference` elements.
There are plenty of reasons why you can end-up having duplicates which are hard to spot:
 - package references defined via `Directory.Build.props` 
 - package references implicitly defined in the SDK (e.g. `FSharp.Core` in `*.fsproj` files)
 - package references defined by imported `*.props` files

Silently accepting these duplicates leads to unnecessary confusion. E.g. I can go to my `*.fsproj` project file and update the `FSharp.Core` version but it will have no effect on the version that is actually used (`FSharp.Core` is by default defined by the sdk unless `DisableImplicitFSharpCoreReference` is specified). Inexperienced developers don't even know how to debug such issues so it can make them feel helpless.

**Comment by davkean** [comment]
I've also spent a bunch diagnosing duplicate package references too. A warning would be extremely helpful.

**Comment by drewnoakes** [comment]
Could this be introduced behind a warning wave?

**Comment by nkolev92** [comment]
Can you elaborate what you mean by a warning wave?

**Comment by drewnoakes** [comment]
Looks like it's called a Change Wave in the docs: https://docs.microsoft.com/en-us/visualstudio/msbuild/change-waves?view=vs-2022

**Comment by nkolev92** [comment]
Hey all, 

Given that this can be a pretty impactful change, I have written up a proposal. 
I would appreciate any reviews on it https://github.com/NuGet/Home/pull/11649. 

The pull request for the proposal can be found at https://github.com/NuGet/NuGet.Client/pull/4484.

**Comment by nkolev92** [comment]
tldr of the proposal:

When a NuGet item is seen as a duplicate, the first item is chosen and a warning is raised. NU1505 for PackageReference, NU1506 for PackageDownload, NU1507 for PackageVersion. 
This warning supports NoWarn, TreatWarningsAsErrors, and it can be disabled altogether.

We're also supporting ContinueOnError for VS purposes.

**Comment by nkolev92** [comment]
Reviewed the proposal internally. 
Goal is 17.3/6.0.400 for VS/.NET SDK respectively
